packages/vite/src/node/server/middlewares/proxy.ts

```diff
@@ -194,24 +200,26 @@ export function proxyMiddleware(
   }
 
   // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
-  return function viteProxyMiddleware(req, res, next) {
+  return async function viteProxyMiddleware(req, res, next) {
     const url = req.url!
     for (const context in proxies) {
       if (doesProxyContextMatchUrl(context, url)) {
         const [proxy, opts] = proxies[context]
         const options: HttpProxy.ServerOptions = {}
 
         if (opts.bypass) {
-          const bypassResult = opts.bypass(req, res, opts)
+          const bypassResult = await opts.bypass(req, res, opts)
```

Comment:
This is the `bypass` call that matters almost all of the time; the upgrade event handling above should probably be consistent, but please apply my comment (and probably all comments) from above to this fn.